File: AlgorithmsPackCPP/1043. Partition Array for Maximum Sum.cpp
const int MAX = 500+1;
int memory[MAX];
vector<int>arr;
int k;

class Solution {
public:
    int partition(int start){
        if(start==arr.size())
            return 0;
        
        auto&ret = memory[start];
        if(ret!=-1)
            return ret;
        
        ret = INT_MIN;
        int mxValue = INT_MIN;
        for(int split = start; split<start+k; split++){
            if(split>=arr.size())
                break;
            mxValue = max(mxValue,arr[split]);
            int len = split-start+1;
            ret = max(ret, len*mxValue+partition(split+1));
        }
        return ret;
    }
    
    int maxSumAfterPartitioning(vector<int>& arr_, int k_) {
        arr = arr_;
        k = k_;
        memset(memory,-1,sizeof(memory));
        return partition(0);
    }
};

File: AlgorithmsPackJavaSolutions/1043. Partition Array for Maximum Sum.java
class Solution {
    private int[] memory;

    public int maxSumAfterPartitioning(int[] arr, int k) {
        int MAX = 500 + 1;
        memory = new int[MAX];
        Arrays.fill(memory, -1);

        return partition(0, arr, k);
    }

    private int partition(int start, int[] arr, int k) {
        if (start == arr.length) {
            return 0;
        }

        if (memory[start] != -1) {
            return memory[start];
        }

        int result = Integer.MIN_VALUE;
        int maxValue = Integer.MIN_VALUE;

        for (int split = start; split < start + k; split++) {
            if (split >= arr.length) {
                break;
            }
            maxValue = Math.max(maxValue, arr[split]);
            int length = split - start + 1;
            result = Math.max(result, length * maxValue + partition(split + 1, arr, k));
        }

        memory[start] = result;
        return result;
    }
}
